node/bft/ledger-service/src/ledger.rs

```diff
@@ -120,13 +125,47 @@ impl<N: Network, C: ConsensusStorage<N>> LedgerService<N> for CoreLedgerService<
 
     /// Returns the block for the given block height.
     fn get_block(&self, height: u32) -> Result<Block<N>> {
-        self.ledger.get_block(height)
+        if let Some(block) = self.block_cache.read().get(&height) {
+            Ok(block.clone())
+        } else {
+            self.ledger.get_block(height)
+        }
     }
 
     /// Returns the blocks in the given block range.
     /// The range is inclusive of the start and exclusive of the end.
     fn get_blocks(&self, heights: Range<u32>) -> Result<Vec<Block<N>>> {
-        self.ledger.get_blocks(heights)
+        // Prepare blocks to collect.
+        let mut blocks = Vec::with_capacity(heights.len());
+        // Determine the lowest height.
+        let start_height = heights.start;
+        // Prepare missing heights to collect.
+        let mut missing_heights = BTreeSet::from_iter(heights.clone());
+        // For each height in the range, check if the block is in the cache.
+        for height in heights {
+            if let Some(block) = self.block_cache.read().get(&height) {
```

Comment:
it would probably be more perf-friendly to keep a read guard for the entire duration of this loop

Reply:
it could also be used to quickly skip the attempt to obtain blocks from the cache in case its 1st and last entries are outside of the requested range